boxtree/tree_build.py

```diff
@@ -37,6 +37,242 @@ class MaxLevelsExceeded(RuntimeError):
     pass
 
 
+# {{{ box-based tree
+
+def resized_array(arr, new_size):
+    """Return a resized copy of the array. The new_size is a scalar which is
+    applied to the last dimension.
+    """
+    old_size = arr.shape[-1]
+    prefix = (slice(None), ) * (arr.ndim - 1)
+    if old_size >= new_size:
+        return arr[prefix + (slice(new_size), )].copy()
+    else:
+        new_shape = list(arr.shape)
+        new_shape[-1] = new_size
+        new_arr = np.zeros(new_shape, arr.dtype)
+        new_arr[prefix + (slice(old_size), )] = arr
+        return new_arr
+
+
+def vec_of_signs(dim, i):
+    """The sign vector is obtained by converting i to a dim-bit binary.
+    """
+    # e.g. bin(10) = '0b1010'
+    binary_digits = [int(bd) for bd in bin(i)[2:]]
+    n = len(binary_digits)
+    assert n <= dim
+    return np.array([0]*(dim-n) + binary_digits) * 2 - 1
+
+
+class _TreeOfBoxes:
+    def __init__(
+            self, box_centers, root_box_extent,
+            box_parents, box_children, box_levels):
+        self.box_centers = box_centers
+        self.root_box_extent = root_box_extent
+        self.box_parents = box_parents
+        self.box_children = box_children
+        self.box_levels = box_levels
+
+    def copy(self):
+        return _TreeOfBoxes(
+                box_centers=self.box_centers.copy(),
+                root_box_extent=self.root_box_extent.copy(),
+                box_parents=self.box_parents.copy(),
+                box_children=self.box_children.copy(),
+                box_levels=self.box_levels.copy())
+
+    @property
+    def dim(self):
+        return self.box_centers.shape[0]
+
+    # {{{ dummy interface for TreePlotter
+
+    @property
+    def dimensions(self):
+        return self.dim
+
+    @property
+    def bounding_box(self):
+        lows = self.box_centers[:, 0] - 0.5*self.root_box_extent
+        highs = lows + self.root_box_extent
+        return [lows, highs]
+
+    def get_box_size(self, ibox):
+        lev = self.box_levels[ibox]
+        box_size = self.root_box_extent * 0.5**lev
+        return box_size
+
+    def get_box_extent(self, ibox):
+        box_size = self.get_box_size(ibox)
+        extent_low = self.box_centers[:, ibox] - 0.5*box_size
+        extent_high = extent_low + box_size
+        return extent_low, extent_high
+
+    # }}} End dummy interface for TreePlotter
+
+    @property
+    def nboxes(self):
+        return self.box_centers.shape[1]
+
+    def nlevels(self):
+        return max(self.box_levels) + 1  # level starts from 0
+
+    def is_leaf(self):
+        # box_id -> whether the box is leaf
+        return np.all(self.box_children == 0, axis=0)
+
+    def leaf_boxes(self):
+        boxes = np.arange(self.nboxes)
+        return boxes[self.is_leaf()]
+
+    # {{{ instantiate with particles
+
+    def distribute_quadrature_rule(self, quad_rule):
```

Comment:
Make a function?

Reply:
Actually, the first half of this function is re-implementing `modepy.TensorProductQuadrature`, but with a specific ordering of nodes. 

[`modepy.nodes.tensor_product_nodes`](https://github.com/inducer/modepy/blob/ee3754fbaa68724593527e83f8c4d337ad080189/modepy/nodes.py#L340) does not seem to stick to an ordering. Currently, it seems to produce nodes in reverse lexicographic order.

Reply:
5170b30d67bf56b6cac9bbf94075ebbc6150d654